### PR TITLE
feat: print etcd ready status

### LINKED ISSUE #3021: When filtering etcd or api-server pods, the readiness status of the pods was not checked.
## What keywords did you search in kubeadm issues before filing this one?

If you have found any duplicates, you should instead reply there and close this page.

If you have not found any duplicates, delete this section and continue on.

## Is this a BUG REPORT or FEATURE REQUEST?

Choose one: BUG REPORT or FEATURE REQUEST
BUG

## Versions

**kubeadm version** (use `kubeadm version`): all

**Environment**:
- **Kubernetes version** (use `kubectl version`):
- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Container runtime (CRI)** (e.g. containerd, cri-o):
- **Container networking plugin (CNI)** (e.g. Calico, Cilium):
- **Others**:

## What happened?
When filtering etcd or api-server pods, the readiness status of the pods was not checked.

## What you expected to happen?
When filtering etcd or api-server pods, the readiness status of the pods was checked.

### PR DESCRIPTION
**Pull request by yxxhero**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
see: https://github.com/kubernetes/kubeadm/issues/3021
#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubeadm/issues/3021

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: cmd/kubeadm/app/util/etcd/etcd.go [hunk 1]
```diff
@@ -229,6 +230,9 @@ func getRawEtcdEndpointsFromPodAnnotationWithoutRetry(client clientset.Interface
 	}
 	etcdEndpoints := []string{}
 	for _, pod := range podList.Items {
+		if !v1pod.IsPodReady(&pod) {
+			klog.V(3).Infof("pod %q is not ready", pod.ObjectMeta.Name)
+		}
```
Comment (neolit123) [inline cmd/kubeadm/app/util/etcd/etcd.go:235]: we cannot import k8s.io/kubernetes/pkg in kubeadm
you can try printing the status another way.

Comment (neolit123) [reply cmd/kubeadm/app/util/etcd/etcd.go:235]: https://github.com/kubernetes/kubernetes/blob/237d3dfda70d898e8b7641a295f5bdfecd527c2e/staging/src/k8s.io/kubectl/pkg/util/podutils/podutils.go#L186

Comment (yxxhero) [reply cmd/kubeadm/app/util/etcd/etcd.go:235]: @neolit123  thanks so much.

**Comment by yxxhero** [comment]
@neolit123 done. thanks so much again.

File: cmd/kubeadm/app/util/staticpod/utils.go [hunk 2]
```diff
@@ -429,3 +429,16 @@ func DeepHashObject(hasher hash.Hash, objectToWrite interface{}) {
 	hasher.Reset()
 	fmt.Fprintf(hasher, "%v", dump.ForHash(objectToWrite))
 }
+
+// IsPodReady checks if a pod is ready.
+// It iterates through the conditions of the pod's status and returns true if it finds a condition of type "PodReady" with status "ConditionTrue".
+// Otherwise, it returns false.
+func IsPodReady(pod *v1.Pod) bool {
```
Comment (neolit123) [inline cmd/kubeadm/app/util/staticpod/utils.go:436]: i don't think we need the new helper function. just use the code inside etcd.go

Comment (yxxhero) [reply cmd/kubeadm/app/util/staticpod/utils.go:436]: @neolit123 ok. got it.

Comment (yxxhero) [reply cmd/kubeadm/app/util/staticpod/utils.go]: @neolit123 done. thanks so much.

File: cmd/kubeadm/app/util/etcd/etcd.go [hunk 3]
```diff
@@ -229,6 +230,16 @@ func getRawEtcdEndpointsFromPodAnnotationWithoutRetry(client clientset.Interface
 	}
 	etcdEndpoints := []string{}
 	for _, pod := range podList.Items {
+		podIsReady := false
+		for _, c := range pod.Status.Conditions {
+			// we only care about pod ready conditions
```
Comment (neolit123) [inline cmd/kubeadm/app/util/etcd/etcd.go:235]:
```suggestion
```

File: cmd/kubeadm/app/util/etcd/etcd.go [hunk 4]
```diff
@@ -229,6 +230,16 @@ func getRawEtcdEndpointsFromPodAnnotationWithoutRetry(client clientset.Interface
...
+			if c.Type == corev1.PodReady && c.Status == corev1.ConditionTrue {
+				podIsReady = true
```
Comment (neolit123) [inline cmd/kubeadm/app/util/etcd/etcd.go:237]:
```suggestion
				podIsReady = true
				break
```

**Comment by yxxhero** [comment]
@neolit123 thanks so much.

**Review by neolit123 — COMMENTED** [review]
/triage accepted
/approve
/priority backlog

**Review by liangyuanpeng — APPROVED** [review]
LGTM

**Pull request merged**